pyccel/ast/core.py

```diff
@@ -4941,32 +4935,51 @@ class Slice(Basic):
     end : Symbol or int
         ending index
 
+    step : Symbol or int default None
+
     Examples
     --------
     >>> from sympy import symbols
     >>> from pyccel.ast.core import Slice
-    >>> m, n = symbols('m, n', integer=True)
-    >>> Slice(m,n)
-    m : n
-    >>> Slice(None,n)
-     : n
-    >>> Slice(m,None)
-    m :
+    >>> start, end, step = symbols('start, end, step', integer=True)
+    >>> Slice(start, end)
+    start : end
+    >>> Slice(None, end)
+     : end
+    >>> Slice(start, None)
+    start :
+    >>> Slice(start, end, step)
+    start : end : step
     """
 
-    # TODO add step
-    # TODO check that args are integers
-    # TODO add negative indices
-    def __new__(cls, start, end):
-        return Basic.__new__(cls, start, end)
+    def __new__(cls, start, end, step = None):
+        return Basic.__new__(cls, start, end, step)
+
+    def __init__(self, start, end, step = None):
+        if start is not None and not (isinstance(start, (Symbol, PyccelOperator)) or
+                (hasattr(start, 'dtype') and isinstance(start.dtype, NativeInteger))):
```

Comment:
I'm not getting the data type of PyccelUnarySub.

Reply:
Why not?

Reply:
```
hasattr(start, 'dtype') = True
start.dtype  = None
```

Reply:
So there is a bug in `PyccelUnarySub`?

I just did:
```
In [11]: from pyccel.ast.core import  Variable

In [13]: a = Variable('int','a')

In [14]: a
Out[14]: a

In [15]: a.shape
Out[15]: ()

In [17]: from pyccel.ast.operators import PyccelUnarySub

In [18]: b = PyccelUnarySub(a)

In [19]: b.dtype
Out[19]: NativeInteger()
```
and it seems to work. So it shouldn't be a bug in `PyccelUnarySub`. Is it possible that you have forgotten to `_visit` something?

Reply:
Never mind, I just realised that this is `ast/core.py` so you will also be checking this in the syntactic stage. In this file you will find lots of lines doing:
```
if self.stage == 'syntactic':
            return
```
This is because the `dtype` isn't known at the syntactic stage. This seems like a better approach as your approach won't work in other unexpected circumstances. Eg.:
```
import numpy as np

def get_step():
    return 1

a = np.ones(10)
b = a[4:7:get_step()]
```